### PR TITLE
Load unload change notification A3IDES-560, A3IDES-669, A3IDES-674

### PR DESCRIPTION
**Pull request by RadekVana**

### CODE REVIEW

File: src/common/dialog_commands.hpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+enum class Command : uint8_t {
+    _NONE = 0, //none must be zero becouse of empty initialization of array
+    YES,
+    NO,
+    CONTINUE,
+    OK,
+    BACK,
+    RETRY,
+    PURGE_MORE,
+    REHEAT
```
Comment (mkbel) [inline src/common/dialog_commands.hpp]: I would prefer to use ALL UPPERCASE only for preprocessor macros.

File: src/common/marlin_server.hpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+//template used by using statement
+template <int VAR_ID, class T>
+class Notifier : public Dialog_notifier {
+public:
+    Notifier(dialog_t type, uint8_t phase, T min, T max, uint8_t progress_min, uint8_t progress_max)
+        : Dialog_notifier(type, phase, cvariant8(min), cvariant8(max), progress_min, progress_max, VAR_ID) {}
+};
+
+//use an alias to automatically notyfi progress
```
Comment (mkbel) [inline src/common/marlin_server.hpp]: notyfi -> notify

File: src/gui/dialogs/DialogLoadUnload.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
+#include "DialogLoadUnload.hpp"
+#include "DialogLoadUnload.h"
+#include "gui.h"      //resource_font
+#include "resource.h" //IDR_FNT_BIG
+
+#ifdef __cplusplus
+extern "C" {
+#endif //__cplusplus
+int16_t WINDOW_CLS_DLG_LOADUNLOAD = 0;
```
Comment (mkbel) [inline src/gui/dialogs/DialogLoadUnload.cpp]: I would prefer to use ALL UPPERCASE only for preprocessor macros.

File: src/gui/dialogs/DialogLoadUnload.hpp [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include "DialogStateful.hpp"
+
+//constexpr size_t DialogLoadUnloadPhases = CountPhases<PhasesLoadUnload>();
```
Comment (mkbel) [inline src/gui/dialogs/DialogLoadUnload.hpp]: Commented out code.

File: src/gui/dialogs/DialogRadioButton.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    button_draw(win.rect, texts[0], win.pfont, IsEnabled());
+}
+
+void RadioButton::draw_n_btn(size_t btn_count) const {
+    rect_ui16_t rc_btn = win.rect;
+
+    int16_t btn_width = rc_btn.w / btn_count - gui_defaults.btn_spacing * (btn_count - 1);
+
+    rc_btn.w = btn_width;
+    //lhs button
```
Comment (mkbel) [inline src/gui/dialogs/DialogRadioButton.cpp]: I don't understand this comment.

File: src/gui/dialogs/DialogRadioButton.cpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+void RadioButton::draw_0_btn() const {
+    display->fill_rect(win.rect, win.color_back);
+}
+
+void RadioButton::draw_1_btn() const {
+    button_draw(win.rect, texts[0], win.pfont, IsEnabled());
+}
+
+void RadioButton::draw_n_btn(size_t btn_count) const {
```
Comment (mkbel) [inline src/gui/dialogs/DialogRadioButton.cpp]: Function name is not very clear. Is it draw n-th button, or draw n buttons?

File: src/gui/dialogs/DialogStateful.cpp [hunk 7]
```diff
@@ -0,0 +1,131 @@
+#include "DialogStateful.hpp"
+#include "DialogRadioButton.hpp"
+#include "gui.h"
+
+extern window_t *window_1; //current popup window, C-code remain
```
Comment (mkbel) [inline src/gui/dialogs/DialogStateful.cpp]: Do we need it?

File: src/gui/dialogs/DialogStateful.cpp [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    return true;
+}
+
+IDialogStateful::~IDialogStateful() {
+    window_destroy(id);
+    window_set_capture(id_capture);
+    window_invalidate(0);
+}
+
+extern window_t *window_1; //current popup window
```
Comment (mkbel) [inline src/gui/dialogs/DialogStateful.cpp]: It is declared twice in the same file.

File: src/gui/dialogs/DialogStateful.cpp [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+void IDialogStateful::draw_frame() {
+    rect_ui16_t rc = rect;
+    display->draw_line(point_ui16(rc.x, rc.y), point_ui16(239, rc.y), COLOR_GRAY);
+    display->draw_line(point_ui16(rc.x, rc.y), point_ui16(rc.x, 320 - 67), COLOR_GRAY);
+    display->draw_line(point_ui16(239, rc.y), point_ui16(239, 320 - 67), COLOR_GRAY);
+    display->draw_line(point_ui16(rc.x, 320 - 67), point_ui16(239, 320 - 67), COLOR_GRAY);
+}
+
+//this should be moved elswhere
```
Comment (mkbel) [inline src/gui/dialogs/DialogStateful.cpp]: Use TODO in such comment, it can be find easily.

File: src/gui/dialogs/DialogStateful.cpp [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    rc_pro.w = w - rc_pro.w;
+    display->fill_rect(rc_pro, COLOR_GRAY);
+    rc_pro.y += rc_pro.h;
+    rc_pro.w = win_rect.w - 120;
+    rc_pro.x = win_rect.x + 60;
+    rc_pro.h = 30;
+    sprintf(text, "%d%%", progress);
+    render_text_align(rc_pro, text, font, color_back, color_text, padding, ALIGN_CENTER);
+}
+//this should be moved elswhere
```
Comment (mkbel) [inline src/gui/dialogs/DialogStateful.cpp]: Use TODO in such comment, it can be find easily.

File: src/gui/dialogs/DialogStateful.hpp [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+#include "DialogRadioButton.hpp"
+#include "marlin_client.hpp"
+#include "dialog_commands.hpp"
+
+#pragma pack(push)
+#pragma pack(1)
+
+//#define DLG_FRAME_ENA 1
+#define DLG_FRAME_ENA 0
+//general foe any number of phases
```
Comment (mkbel) [inline src/gui/dialogs/DialogStateful.hpp]: General enemy?

File: src/gui/dialogs/DialogStateful.hpp [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+        DLG_DRA_FR = 0x0000, // draw frame
+#endif                                          //DLG_FRAME_ENA == 1
+        DLG_TXT_CH = 0x2000,                    // text changed
+        DLG_PRO_CH = 0x4000,                    // progress changed
+        DLG_PPR_CH = 0x8000,                    // part progress changed
+        DLG_PRX_CH = (DLG_PRO_CH | DLG_PPR_CH), // some progress changed
+        DLG_PHA_CH = (DLG_PRX_CH | DLG_TXT_CH), // phase changed
+        //dialog flags bitmasks
+        DLG_MSK_MOD = 0x0003,    // mode mask
+        DLG_MSK_CHG = DLG_PHA_CH // change flag mask
```
Comment (mkbel) [inline src/gui/dialogs/DialogStateful.hpp]: I would prefer to use ALL UPPERCASE only for preprocessor macros.

File: src/common/dialog_commands.hpp (hunk 1 above)
Comment (RadekVana) [reply src/common/dialog_commands.hpp]: done

File: src/gui/dialogs/DialogLoadUnload.cpp (hunk 3 above)
Comment (RadekVana) [reply src/gui/dialogs/DialogLoadUnload.cpp]: all window class IDs has this format

File: src/common/marlin_server.hpp [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+using Notifier_PRNSPEED = Notifier<MARLIN_VAR_PRNSPEED, uint16_t>;
+using Notifier_FLOWFACT = Notifier<MARLIN_VAR_FLOWFACT, uint16_t>;
+using Notifier_WAITHEAT = Notifier<MARLIN_VAR_WAITHEAT, uint8_t>;
+using Notifier_WAITUSER = Notifier<MARLIN_VAR_WAITUSER, uint8_t>;
+using Notifier_SD_PRINT = Notifier<MARLIN_VAR_SD_PRINT, uint8_t>;
+using Notifier_SD_PDONE = Notifier<MARLIN_VAR_SD_PDONE, uint8_t>;
+using Notifier_DURATION = Notifier<MARLIN_VAR_DURATION, uint32_t>;
+
+//create dialog and automatically destroy it at the end of scope
+class DialogRAII {
```
Comment (mkbel) [inline src/common/marlin_server.hpp]: Wouln't it be possible to find better name for this class? For example just Dialog or Dialog_opener or Dialog_helper?

File: src/common/marlin_server.hpp [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+    dialog_t dialog;
+
+public:
+    DialogRAII(dialog_t type, uint8_t data)
```
Comment (mkbel) [inline src/common/marlin_server.hpp]: What is data? I am missing documentation or self explaining parameter name.

File: src/gui/dialogs/DialogStateful.cpp (hunk 7 above)
Comment (RadekVana) [reply src/gui/dialogs/DialogStateful.cpp]: now - yes 
in future - no

File: src/common/marlin_server.hpp (hunk 13 above)
Comment (mkbel) [reply src/common/marlin_server.hpp]: Or DialogOpener or DialogHelper, I don't know what naming convention we are using.

File: src/gui/dialogs/DialogStateful.hpp (hunk 12 above)
Comment (RadekVana) [reply src/gui/dialogs/DialogStateful.hpp]: this "flags" are used like preprocessor macros
that enum has no name
enum is used only to enclose those "macros" within class scope
also it is just moved old code

File: src/common/marlin_server.hpp (hunk 2 above)
Comment (RadekVana) [reply src/common/marlin_server.hpp]: fixed

File: src/gui/dialogs/DialogLoadUnload.hpp (hunk 4 above)
Comment (RadekVana) [reply src/gui/dialogs/DialogLoadUnload.hpp]: removed

File: src/gui/dialogs/DialogRadioButton.cpp (hunk 5 above)
Comment (RadekVana) [reply src/gui/dialogs/DialogRadioButton.cpp]: copy - paste misstake
removed

File: src/gui/dialogs/DialogRadioButton.cpp (hunk 6 above)
Comment (RadekVana) [reply src/gui/dialogs/DialogRadioButton.cpp]: renamed to draw_n_btns

File: src/gui/dialogs/DialogStateful.cpp (hunk 9 above)
Comment (RadekVana) [reply src/gui/dialogs/DialogStateful.cpp]: fixed

File: src/gui/dialogs/DialogStateful.cpp (hunk 10 above)
Comment (RadekVana) [reply src/gui/dialogs/DialogStateful.cpp]: fixed

File: src/gui/dialogs/DialogStateful.hpp (hunk 11 above)
Comment (RadekVana) [reply src/gui/dialogs/DialogStateful.hpp]: fixed

File: src/common/marlin_server.hpp (hunk 13 above)
Comment (RadekVana) [reply src/common/marlin_server.hpp]: it is "to be renamed"
i see bigger problem in "Dialog part of name", marlin_server should not care about dialogs

File: src/common/marlin_server.hpp (hunk 14 above)
Comment (RadekVana) [reply src/common/marlin_server.hpp]: added comment

File: src/gui/dialogs/DialogStateful.cpp (hunk 8 above)
Comment (RadekVana) [reply src/gui/dialogs/DialogStateful.cpp]: removed

File: src/common/marlin_server.hpp (hunk 13 above)
Comment (RadekVana) [reply src/common/marlin_server.hpp]: renamed

File: src/common/marlin_server.cpp [hunk 15]
```diff
@@ -1,7 +1,7 @@
 // marlin_server.cpp
 
 #include "marlin_server.h"
```
Comment (joshymjose) [inline src/common/marlin_server.cpp]: still valid .h?

Comment (RadekVana) [reply src/common/marlin_server.cpp]: yes

Comment (RadekVana) [reply src/common/marlin_server.cpp]: it is C-interface

File: src/marlin_stubs/pause/pause.cpp [hunk 16]
```diff
@@ -115,33 +119,64 @@ static void filament_change_beep(const int8_t max_beep_count, const bool init =
...
-    #endif
     current_position.e += length / planner.e_factor[active_extruder];
     line_to_current_position(fr_mm_s);
     planner.synchronize();
 }
 
+void do_pause_e_move_notify_progress(const float &length, const feedRate_t &fr_mm_s, PhasesLoadUnload phase, uint8_t progress_min, uint8_t progress_max) {
+    //Not sure if folowing code would not be better
+    //const float actual_e = planner.get_axis_position_mm(E_AXIS);
+    //Notifier_POS_E N(ClinetFSM::Load_unload, GetPhaseIndex(phase), actual_e, actual_e + length, progress_min,progress_max);
```
Comment (mkbel) [inline src/marlin_stubs/pause/pause.cpp]: Commented out code.

File: src/marlin_stubs/pause/pause.cpp [hunk 17]
```diff
@@ -115,33 +119,64 @@ static void filament_change_beep(const int8_t max_beep_count, const bool init =
...
+    const float actual_e = current_position.e;
+    current_position.e += length / planner.e_factor[active_extruder];
+    Notifier_POS_E N(ClinetFSM::Load_unload, GetPhaseIndex(phase), actual_e, current_position.e, progress_min, progress_max);
+    line_to_current_position(fr_mm_s);
+    planner.synchronize();
+}
+
+void plan_pause_e_move(const float &length, const feedRate_t &fr_mm_s) {
+    current_position.e += length / planner.e_factor[active_extruder];
+    //line_to_current_position(fr_mm_s);
```
Comment (mkbel) [inline src/marlin_stubs/pause/pause.cpp]: Commented out code.

File: src/marlin_stubs/pause/pause.cpp [hunk 18]
```diff
@@ -115,33 +119,64 @@ static void filament_change_beep(const int8_t max_beep_count, const bool init =
...
+    while (!planner.buffer_line(current_position, fr_mm_s, active_extruder)) {
+        delay(50);
+    }
+}
+
+void plan_pause_e_move_notify_progress(const float &length, const feedRate_t &fr_mm_s, PhasesLoadUnload phase, uint8_t progress_min, uint8_t progress_max) {
+    const float actual_e = current_position.e;
+    current_position.e += length / planner.e_factor[active_extruder];
+    Notifier_POS_E N(ClinetFSM::Load_unload, GetPhaseIndex(phase), actual_e, current_position.e, progress_min, progress_max);
+    //line_to_current_position(fr_mm_s);
```
Comment (mkbel) [inline src/marlin_stubs/pause/pause.cpp]: Commented out code.

File: src/marlin_stubs/pause/pause.cpp [hunk 19]
```diff
@@ -158,163 +193,46 @@ bool load_filament(const float &slow_load_length /*=0*/, const float &fast_load_
...
+        idle(true);
 
-    #if ENABLED(DUAL_X_CARRIAGE)
-    const int8_t saved_ext = active_extruder;
-    const bool saved_ext_dup_mode = extruder_duplication_enabled;
-    active_extruder = DXC_ext;
-    extruder_duplication_enabled = false;
-    #endif
+    //check FILAMET SENSOR
+    //todo
```
Comment (mkbel) [inline src/marlin_stubs/pause/pause.cpp]: Ambiguous todo.

File: src/marlin_stubs/pause/pause.cpp [hunk 20]
```diff
@@ -24,47 +24,48 @@
...
-
-    #include "../../lib/Marlin/Marlin/src/libs/nozzle.h"
-    #include "../../lib/Marlin/Marlin/src/feature/pause.h"
+#include "../../lib/Marlin/Marlin/src/libs/nozzle.h"
+#include "../../lib/Marlin/Marlin/src/feature/pause.h"
 
+#include "marlin_server.hpp"
 // private:
-
+//check unsupported feacures
```
Comment (mkbel) [inline src/marlin_stubs/pause/pause.cpp]: Typo feacures -> fuck-cures or features

File: src/marlin_stubs/pause/pause.cpp [hunk 21]
```diff
@@ -24,47 +24,48 @@
...
+//filament sensor is no longre part of marlin thus it must be disabled
```
Comment (mkbel) [inline src/marlin_stubs/pause/pause.cpp]: Typo longre -> KFC Longer

File: src/marlin_stubs/pause/pause.cpp (hunk 16 above)
Comment (RadekVana) [reply src/marlin_stubs/pause/pause.cpp]: I plan to keep this comment for now

File: src/marlin_stubs/pause/pause.cpp (hunk 17 above)
Comment (RadekVana) [reply src/marlin_stubs/pause/pause.cpp]: removed

File: src/marlin_stubs/pause/pause.cpp (hunk 18 above)
Comment (RadekVana) [reply src/marlin_stubs/pause/pause.cpp]: removed

File: src/marlin_stubs/pause/pause.cpp (hunk 21 above)
Comment (RadekVana) [reply src/marlin_stubs/pause/pause.cpp]: thx

File: src/marlin_stubs/pause/pause.cpp (hunk 19 above)
Comment (RadekVana) [reply src/marlin_stubs/pause/pause.cpp]: there is missing implementation of filament sensor state handling
merged 2 comments in one

File: src/common/client_response.hpp [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+    static const PhaseResponses TestResponses[CountPhases<PhasesTest>()];
+
+    //methods to "bind" button array with enum type
+    static const PhaseResponses &getResponsesInPhase(PhasesLoadUnload phase) { return LoadUnloadResponses[static_cast<size_t>(phase)]; }
+    static const PhaseResponses &getResponsesInPhase(PhasesTest phase) { return TestResponses[static_cast<size_t>(phase)]; }
+
+protected:
+    //get index of single response in PhaseResponses
+    template <class T>
+    static uint8_t GetIndex(T phase, Response response) {
```
Comment (mkbel) [inline src/common/client_response.hpp]: Return value not documented. Type of return value is unsigned, but we return negative number.

File: src/common/client_response.hpp [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+// buttons are generalized on this level as responses
+// because non GUI/WUI client can also use them
+
+#pragma once
+
+#include <cstdint>
+#include <cstddef>
+#include <array>
+enum { RESPONSE_BITS = 2,
+    MAX_RESPONSES = (1 << RESPONSE_BITS) };
```
Comment (mkbel) [inline src/common/client_response.hpp]: Not documented, meaning not clear.

File: src/common/client_response.hpp [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+
+//list of all button types
+enum class Response : uint8_t {
+    _none = 0, //none must be zero becouse of empty initialization of array
```
Comment (mkbel) [inline src/common/client_response.hpp]: Typo becouse -> because

File: src/common/client_response.hpp [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+
+using PhaseResponses = std::array<Response, MAX_RESPONSES>;
+
+//count cenum class members (if "_first" and "_last" is defined)
+template <class T>
+constexpr size_t CountPhases() {
+    return static_cast<size_t>(T::_last) - static_cast<size_t>(T::_first) + 1;
+}
+//use this when creating an event
+//encodes enum to position in Enem
```
Comment (mkbel) [inline src/common/client_response.hpp]: typo Enem -> Enema

File: src/common/client_response.hpp [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+    //get all responses accepted in phase
+    template <class T>
+    static const PhaseResponses &GetResponses(T phase) {
+        return getResponsesInPhase(phase);
+    }
+
+    //encode radio button and clicked index into int
+    //use on client side
+    template <class T>
+    static uint32_t Encode(T phase, Response response) {
```
Comment (mkbel) [inline src/common/client_response.hpp]: Return value is of unsigned type, but we are returning negative number.

File: src/common/marlin_client.c [hunk 27]
```diff
@@ -154,24 +155,32 @@ int marlin_client_id(void) {
     return 0;
 }
 
-int marlin_client_set_dialog_open_cb(dialog_open_cb_t cb) {
+int marlin_client_set_fsm_create_cb(fsm_create_t cb) {
```
Comment (mkbel) [inline src/common/marlin_client.c]: I have no clue what cb is. Return value is not documented.

Comment (RadekVana) [reply src/common/marlin_client.c]: callback

File: src/common/marlin_client.c [hunk 28]
```diff
@@ -154,24 +155,32 @@ int marlin_client_id(void) {
...
     if (client && cb) {
-        client->dialog_open_cb = cb;
+        client->fsm_create_cb = cb;
         return 1;
     }
     return 0;
 }
 
-int marlin_client_set_dialog_close_cb(dialog_close_cb_t cb) {
+int marlin_client_set_fsm_destroy_cb(fsm_destroy_t cb) {
```
Comment (mkbel) [inline src/common/marlin_client.c]: I have no clue what cb is. Return value is not documented.

File: src/common/marlin_client.c [hunk 29]
```diff
@@ -154,24 +155,32 @@ int marlin_client_id(void) {
...
     marlin_client_t *client = _client_ptr();
     if (client && cb) {
-        client->dialog_close_cb = cb;
+        client->fsm_destroy_cb = cb;
         return 1;
     }
     return 0;
 }
 
+int marlin_client_set_fsm_change_cb(fsm_change_t cb) {
```
Comment (mkbel) [inline src/common/marlin_client.c]: I have no clue what cb is. Return value is not documented.

File: src/common/marlin_client.c [hunk 30]
```diff
@@ -624,6 +633,18 @@ int marlin_reheating(void) {
     return 0;
 }
 
+//-----------------------------------------------------------------------------
+// responses fromclient finite state machine (like button click)
```
Comment (mkbel) [inline src/common/marlin_client.c]: typo fromclient -> from client

File: src/common/marlin_server.cpp [hunk 31]
```diff
@@ -1299,3 +1299,57 @@ void host_prompt_do(const PromptReason type, const char *const pstr, const char
...
+    s_data.progress_max = progress_max;
+    s_data.var_id = var_id;
+    s_data.last_progress_sent = -1;
+}
+
+//static method
+//x = (actual - s_data.min) * s_data.scale + s_data.progress_min;
+//x = actual * s_data.scale - s_data.min * s_data.scale + s_data.progress_min;
+// s_data.offset == -s_data.min * s_data.scale + s_data.progress_min
+//x = actual * s_data.scale + s_data.offset;
```
Comment (mkbel) [inline src/common/marlin_server.cpp]: This comment is not clear for me.

File: src/common/marlin_server.cpp [hunk 32]
```diff
@@ -1299,3 +1299,57 @@ void host_prompt_do(const PromptReason type, const char *const pstr, const char
...
+    if (s_data.type == ClinetFSM::_none)
+        return;
+
+    cvariant8 temp;
+    temp.attach(marlin_vars_get_var(&(marlin_server.vars), s_data.var_id));
+
+    float actual = static_cast<float>(temp);
+    actual = actual * s_data.scale + s_data.offset;
+
+    int progress = int(actual); //int - must be signed
```
Comment (mkbel) [inline src/common/marlin_server.cpp]: c style cast in cpp file.

File: src/common/marlin_server.cpp [hunk 33]
```diff
@@ -1299,3 +1299,57 @@ void host_prompt_do(const PromptReason type, const char *const pstr, const char
...
+}
+
+FSM_notifier::~FSM_notifier() {
+    s_data = temp_data;
+}
+
+/*****************************************************************************/
+//ClientResponseHandler
+//define static member
+uint32_t ClientResponseHandler::server_side_encoded_response = -1;
```
Comment (mkbel) [inline src/common/marlin_server.cpp]: Please document meaning of negative value assigned to unsigned type.

File: src/common/marlin_server.hpp [hunk 34]
```diff
@@ -0,0 +1,122 @@
...
+#include "client_response.hpp"
+
+/*****************************************************************************/
+//C++ only features
+
+//notyfy all clients to create finit statemachine, must match fsm_create_t signature
+void fsm_create(ClinetFSM type, uint8_t data);
+//notyfy all clients to destroy finit statemachine, must match fsm_destroy_t signature
+void fsm_destroy(ClinetFSM type);
+//notyfy all clients to change state of finit statemachine, must match fsm_change_t signature
```
Comment (mkbel) [inline src/common/marlin_server.hpp]: Typo notyfy -> notify. Put TODO for must match fsm_change_t signature comment.

File: src/common/marlin_server.hpp [hunk 35]
```diff
@@ -0,0 +1,122 @@
...
+void fsm_change(ClinetFSM type, uint8_t phase, uint8_t progress_tot, uint8_t progress);
+
+//inheritred class for server side to be able to work with server_side_encoded_response
```
Comment (mkbel) [inline src/common/marlin_server.hpp]: Typo inheritred -> inherited

File: src/common/marlin_server.hpp [hunk 36]
```diff
@@ -0,0 +1,122 @@
...
+            , phase(0)
+            , var_id(0)
+            , last_progress_sent(-1) {}
+    };
+    //static members
+    static data s_data;
+
+    //temporary members
+    //to store previous state
+    data temp_data;
```
Comment (mkbel) [inline src/common/marlin_server.hpp]: Meaning not very clear.

File: src/common/client_response.hpp (hunk 22 above)
Comment (RadekVana) [reply src/common/client_response.hpp]: documented

File: src/common/client_response.hpp (hunk 23 above)
Comment (RadekVana) [reply src/common/client_response.hpp]: documented

File: src/common/client_response.hpp (hunk 24 above)
Comment (RadekVana) [reply src/common/client_response.hpp]: fixed

File: src/common/client_response.hpp (hunk 25 above)
Comment (RadekVana) [reply src/common/client_response.hpp]: fixed

File: src/common/client_response.hpp (hunk 26 above)
Comment (RadekVana) [reply src/common/client_response.hpp]: fixed

File: src/common/marlin_client.c (hunk 27 above)
Comment (RadekVana) [reply src/common/marlin_client.c]: commented

File: src/common/marlin_client.c (hunk 28 above)
Comment (RadekVana) [reply src/common/marlin_client.c]: commented

File: src/common/marlin_client.c (hunk 29 above)
Comment (RadekVana) [reply src/common/marlin_client.c]: commented

File: src/common/marlin_client.c (hunk 30 above)
Comment (RadekVana) [reply src/common/marlin_client.c]: fixed

File: src/common/marlin_server.cpp (hunk 31 above)
Comment (RadekVana) [reply src/common/marlin_server.cpp]: fixed

File: src/common/marlin_server.cpp (hunk 32 above)
Comment (RadekVana) [reply src/common/marlin_server.cpp]: fixed

File: src/common/marlin_server.cpp (hunk 33 above)
Comment (RadekVana) [reply src/common/marlin_server.cpp]: done

File: src/common/marlin_server.hpp (hunk 34 above)
Comment (RadekVana) [reply src/common/marlin_server.hpp]: done

File: src/common/marlin_server.hpp (hunk 35 above)
Comment (RadekVana) [reply src/common/marlin_server.hpp]: fixed

File: src/common/marlin_server.hpp (hunk 36 above)
Comment (RadekVana) [reply src/common/marlin_server.hpp]: added better comment

File: src/gui/screen_menu_filament.c [hunk 37]
```diff
@@ -109,7 +107,7 @@ int screen_menu_filament_event(screen_t *screen, window_t *window, uint8_t event
                 break;
             case MI_PURGE:
                 p_window_header_set_text(&(psmd->header), "PURGE FILAM.");
-                gui_dlg_purge();
+                //   gui_dlg_purge();
```
Comment (mkbel) [inline src/gui/screen_menu_filament.c]: Commented out code.

File: src/gui/dialogs/DialogStateful.hpp [hunk 38]
```diff
@@ -0,0 +1,185 @@
...
+        }
+    }
+}
+
+template <class T>
+void DialogStateful<T>::event(uint8_t event, void *param) {
+    RadioButton &radio = states[phase].button;
+    switch (event) {
+    case WINDOW_EVENT_BTN_DN:
+    //case WINDOW_EVENT_BTN_UP:
```
Comment (mkbel) [inline src/gui/dialogs/DialogStateful.hpp]: Commented out code.

File: src/gui/screen_menu_filament.c (hunk 37 above)
Comment (RadekVana) [reply src/gui/screen_menu_filament.c]: should be todo, will rewrite soon

File: src/gui/dialogs/DialogStateful.hpp (hunk 38 above)
Comment (RadekVana) [reply src/gui/dialogs/DialogStateful.hpp]: will remove it in next PR

**Pull request merged**